Introducing TS vue-tsc

RedGaming4 opened the pull request:
## Description

Introduced vue-tsc as well as created a RawMonster interface and updated the mapMonsterFromAPI function for stronger type interference to avoid TS from throwing type errors. Moreover, added a generic type for findPaginated so you can define what you expect or default to Record<string, unknown> for reverse compatibility. 

RawMonster Should help with visibility for what type of data is coming from the API.

## Related Issue

*Closes https://github.com/open5e/open5e/issues/714*

## How was this tested?
I ran npm run test and it passed all tests.
As well as npm run dev runs fine.

***Draft PR***

RedGaming4 committed:
Introducing TS Linting

calumbell reviewed (changes requested):
Hey, thanks for your first contribution to the repo!

Spinning it up on local it looks like `npx vue-tsc` is running as intended. Lots of type errors remaining, but these should probably be parked until we have resolved issue #725, which will better support us in getting these TS gremlins fixed.

I have one small change to request (sorry!). The `mapMonsterFromAPI` function in the `EncounterBuilderMonsterSearch.vue` component could use a refactor:

```
  if (monster.document !== undefined) {
    return {
      ...base,
      document: {
        name: monster.document.name,
        key: monster.document.key,
      },
    };
  }
  return base;
```

Assuming the `monster.document !== undefined` condition is checking for the presence of the`document` key on the `monster` object, then it can be refactored as: 

```
if (monster.document) {
  // ...
}
```

Which looks much, much nicer IMO.

It is also a bit difficult to tell where exactly the if-statement ends because of the three sequential closing curly-braces the preceed returning the base case. Would it be possible to get this refactored to improve legibility? Perhaps a ternary operator would be helpful here:

```
return !monster.document
  ? base
  : { 
    ...base,
    document: {
      name: monster.document.name,
      key: monster.document.key,
    }
  };
```

(N.B. I have not tested the above code, treat it as pseudo-code)

Brilliant work! Thanks again for your contribution. Let me know if you need any help getting these changes made :)

RedGaming4 commented:
Completely agree with the changes you’ve highlighted and I’m actually a bit relieved it’s not about the core logic! I’ll work on that quick fix this week. And no worries at all about the request I'm here to learn, so feel free to be as detailed or "**harsh**" as needed.

RedGaming4 committed:
Refactored if statement to ternary operator

RedGaming4 commented:
Refactored the if statement into a conditional (ternary) operator for improved readability. Additionally, I added safeguards to ensure that if name or key are null or undefined, they default to empty strings. Previously, I was only checking if document existed, but not handling the case where name or key could be missing or falsy. Thanks for feedback!

calumbell reviewed (approved):
Thanks for making those changes @RedGaming4! Everything looks good to me on my end. I'll get this approved and merged to `staging`.

Thanks again for your contribution! I am planning to write a whole heap of new issues this week, so keep your eyes on the repo if you want working on something else!

calumbell commented:
(Closing and re-opening PR to re-run validation)

RedGaming4 commented:
> Thanks for making those changes @RedGaming4! Everything looks good to me on my end. I'll get this approved and merged to `staging`.
> 
> Thanks again for your contribution! I am planning to write a whole heap of new issues this week, so keep your eyes on the repo if you want working on something else!

Will keep an eye out! have a couple minor TS Errors fixes i've done missing imports and what nots. And I will go ahead and see if i can work on issue 776 (api) & 731 (website) these upcoming weeks.

calumbell commented:
@RedGaming4 Oh! You got the missing import type-errors fixed? Was about to open an issue for that.  Nice! Open a PR when you get a chance
